vortex-ipc/src/reader.rs

```diff
@@ -135,84 +137,121 @@ impl<'a, R: Read> StreamArrayReader<'a, R> {
+                .unwrap_or_default()
+            {
+                vortex_bail!("Indices must be sorted to take from IPC stream")
+            }
 
-        if !indices.dtype().is_int() {
-            vortex_bail!("Indices must be integers")
-        }
-        if indices.dtype().is_signed_int()
-            && indices
+            if indices
                 .statistics()
-                // min cast should be safe
-                .compute_as_cast::<i64>(Stat::Min)
-                .unwrap()
-                < 0
-        {
-            vortex_bail!("Indices must be positive")
+                .compute_as_cast::<u64>(Stat::NullCount)
+                .unwrap_or_default()
+                > 0
+            {
+                vortex_bail!("Indices must not contain nulls")
+            }
+
+            if !indices.dtype().is_int() {
+                vortex_bail!("Indices must be integers")
+            }
+            if indices.dtype().is_signed_int()
+                && indices
+                    .statistics()
+                    // min cast should be safe
+                    .compute_as_cast::<i64>(Stat::Min)
+                    .unwrap()
+                    < 0
+            {
+                vortex_bail!("Indices must be positive")
+            }
         }
 
         if self.row_offset != 0 {
             vortex_bail!("Stream has already been (at least partially) consumed")
         }
 
-        let mut chunks = Vec::new();
+        Ok(TakeIterator {
+            reader: self,
+            indices,
+            row_offset: 0,
+        })
+    }
+}
 
-        let mut row_offset = self.row_offset;
+pub struct TakeIterator<'a, R: Read> {
+    reader: StreamArrayReader<'a, R>,
+    indices: &'a Array<'a>,
+    row_offset: usize,
+}
 
-        // Continue reading batches from the stream until we run out
-        while let Some(batch) = self.next()? {
-            let left =
-                search_sorted::<usize>(indices, row_offset, SearchSortedSide::Left)?.to_index();
-            let right =
-                search_sorted::<usize>(indices, row_offset + batch.len(), SearchSortedSide::Left)?
-                    .to_index();
-            // NB: if we're done, make sure to consume the rest of the
-            //  array to leave the iterator in a good state
-            if left == indices.len() {
-                continue;
+/// NB: if the StreamArrayReader expires without being fully-consumed by the underlying iterator,
+/// it will leave the underlying messages stream in an unreadable state and free up a mutable ref
+/// to that stream, allowing someone else to try to issue a read that will inevitably fail.
+/// For this reason, we force full consumption of the reader when it goes out of scope.
+impl<'a, R: Read> Drop for StreamArrayReader<'a, R> {
+    fn drop(&mut self) {
+        // NB: can't catch_unwind/unwrap here because &mut self can't be made UnwindSafe
+        loop {
+            let next = self.next();
```

Comment:
if we'd previously reached the end of the array in the StreamArrayReader, in which case `self.next()` previously returned `None`, will it return `None` a second time? (I can check later tonight, fiddly right now on my phone)

Reply:
Yep, added some verification [here](https://github.com/spiraldb/vortex/pull/271/files#diff-ed7229df961bb3810fc5b331b1a2b0258e646c71d47c56c186434590cb7572bfR623)